README.md
# 🔥 mae [![](https://img.shields.io/pypi/v/mae.svg)](https://pypi.org/project/mae) [![](https://img.shields.io/pypi/pyversions/mae.svg)](https://pypi.org/project/mae) [![](https://travis-ci.org/paambaati/mae.svg?branch=master)](https://travis-ci.org/paambaati/mae) [![Maintainability](https://api.codeclimate.com/v1/badges/87b24e61fd1c6a998254/maintainability)](https://codeclimate.com/github/paambaati/mae/maintainability) [![Test Coverage](https://api.codeclimate.com/v1/badges/87b24e61fd1c6a998254/test_coverage)](https://codeclimate.com/github/paambaati/mae/test_coverage)

`mae` collects Prometheus metrics from your Mesos apps.

See [Deployment Strategy](#deployment-strategy) to learn how to set up `mae` on your Mesos cluster.

## Requirements

`mae` requires Python (>= 2.7), and supports Python 3 (>= 3.5).

## Installation

```
pip install mae
```

## Usage
```
$ mae --help

usage: mae [-h] app_port slave_address slave_port

positional arguments:
  app_port       Port on which this exporter will run on
  slave_address  Mesos slave address
  slave_port     Mesos slave port
```
The logging level of the CLI can also be configured with the `LOG_LEVEL` environment variable. Read the [`logging` library's levels](https://docs.python.org/2/library/logging.html#logging-levels) for all the available levels.

## Task Labels

Once `mae` is up and running, it will start collecting metrics from all apps that have the `prometheus.metrics.enabled` label. You can also customize how the metrics are collected  —

| [Task Label](https://docs.mesosphere.com/1.7/usage/tutorials/task-labels/)                           	| Description                                                                                                              	| Required? 	| Default    	|
|---------------------------------	|--------------------------------------------------------------------------------------------------------------------------	|-----------	|------------	|
| `prometheus.metrics.enabled`    	| Enables metrics collection. If the label isn't found, the app's metrics are not collected.                               	| Yes       	|            	|
| `prometheus.metrics.port_index` 	| The port index where your app is exposing its Prometheus metrics. This is useful for Mesos apps that use multiple ports. 	| No        	| `0`        	|
| `prometheus.metrics.endpoint`   	| The endpoint where your app is exposing its metrics.                                                                     	| No        	| `/metrics` 	|

## Deployment Strategy

`mae` is designed to be run as a daemon on all Mesos slave nodes. This ideally involves 2 steps —

1. [Install `mae`](#installation) as part of your base image/AMI or via [user data](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html).
2. Run `mae` as a daemon. For example, if your distro supports `systemd`, here's a sample script that runs the exporter on port `8888` —
    ```
    # Save this to /etc/systemd/system/mae.service
    [Unit]
    Description=Mesos App Exporter
    After=network.target

    [Service]
    Type=simple
    Restart=on-failure
    Environment="LOG_LEVEL=INFO"
    ExecStart=/usr/local/bin/mae 8888 localhost 5051 # Assuming the Mesos slave process is running on port 5051

    [Install]
    WantedBy=multi-user.target
    ```


CHANGELOG.md
# Mae Changelog

`mae` follows [Semantic Versioning](http://semver.org/).

## [1.0.11] - 2019-07-15
### Fixed
- 🐞 Fixed code coverage execution on CI/CI builds.

## [1.0.10] - 2019-07-15
### Added
- Add code coverage.

## [1.0.9] - 2019-07-14
### Changed
- Bump `requests` and `mock` dependencies to latest versions. Thanks @dependabot!

## [1.0.8] - 2019-02-28
### Fixed
- 🐞 Return 200 for when no scrape targets are found, so Prometheus can pick it up.

## [1.0.7] - 2019-02-28
### Fixed
- 🐞 Fixed a bug ([#1](https://github.com/paambaati/mae/issues/1)) where an app without labels can cause an empty response.

## [1.0.6] - 2018-12-31
### Added
- Tests for CLI.

## [1.0.5] - 2018-12-31
### Added
- Type hints 🎉
### Fixed
- 🐞 Python3 import errors - via 39b4a7d.

## [1.0.4] - 2018-12-27
### Added
- Included documentation about Mesos/Marathon task labels.

## [1.0.3] - 2018-12-27
### Changed
- Updated Pydoc.
### Fixed
- CI config fixes.

## [1.0.2] - 2018-12-27
### Added
- `mae` CLI.

## [1.0.1] - 2018-12-27
### Added
- Travis CI builds.

## [1.0.0] - 2018-12-27
### Added
- Initial release.


tests/test_cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
from threading import Thread
from time import sleep
import unittest

THIS_DIR = os.path.dirname(os.path.abspath(__file__))

sys.path.insert(0, os.path.dirname(THIS_DIR))
from mae.cli import ExporterServer


class TestCLIServer(unittest.TestCase):

    def setUp(self):
        self.exporter = ExporterServer(0, 'localhost', 0)

    def tearDown(self):
        self.exporter.socket.close()

    def test_mae_cli_1_simple_start(self):
        """
        Test if CLI can just run and start the Exporter server.
        """
        server_thread = Thread(target=self.exporter.serve_forever)
        server_thread.start()
        sleep(5)
        self.assertTrue(server_thread.is_alive)
        self.exporter.shutdown()
        self.exporter.socket.close()

if __name__ == '__main__':
    unittest.main(verbosity=2, failfast=True)


tests/test_mae.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import json
from mock import Mock, patch
import os
import requests
import sys
import unittest

THIS_DIR = os.path.dirname(os.path.abspath(__file__))

sys.path.insert(0, os.path.dirname(THIS_DIR))
from mae.mae import MesosAppExporter


class TestMesosAppExporter(unittest.TestCase):
    def __readFile(self, filename):
        with open('{0}/{1}'.format(THIS_DIR, filename)) as file:
            data = file.read()
        return data

    def __readJsonFile(self, filename):
        return json.loads(self.__readFile(filename))

    def __mock_response(self, status=200, content='CONTENT', json_data=None, raise_for_status=None):
        mock_resp = Mock()
        mock_resp.raise_for_status = Mock()
        if raise_for_status:
            mock_resp.raise_for_status.side_effect = raise_for_status
        mock_resp.status_code = status
        mock_resp.content = content
        mock_resp.text = content
        if json_data:
            mock_resp.json = Mock(
                return_value=json_data
            )
        return mock_resp

    def setUp(self):
        self.exporter = MesosAppExporter('localhost', 5051)

    def tearDown(self):
        pass

    def test_mae_get_endpoint_1_single_task(self):
        """
        Test if task data with a single matching app is picked up.
        """
        task_data = self.__readJsonFile('fixtures/task_data_1.json')
        endpoints = self.exporter.get_app_metrics_endpoints(task_data)
        self.assertEqual(len(endpoints), 1)
        self.assertEqual(endpoints[0], 'http://localhost:31065/metrics')

    def test_mae_get_endpoint_2_multiple_tasks(self):
        """
        Test if task data with 2 matching apps are picked up.
        Also test if both default and custom labels (endpoint and port index) work.
        """
        task_data = self.__readJsonFile('fixtures/task_data_2.json')
        endpoints = self.exporter.get_app_metrics_endpoints(task_data)
        self.assertEqual(len(endpoints), 2)
        self.assertEqual(endpoints[0], 'http://localhost:31521/metrics')
        self.assertEqual(endpoints[1], 'http://localhost:31065/prometheus')

    def test_mae_get_endpoint_3_no_matching_tasks(self):
        """
        Test if task data without any matching apps are correctly parsed.
        """
        task_data = self.__readJsonFile('fixtures/task_data_3.json')
        endpoints = self.exporter.get_app_metrics_endpoints(task_data)
        self.assertEqual(len(endpoints), 0)

    def test_mae_get_endpoint_4_no_labels(self):
        """
        Test if task data without any labels do not throw any errors.
        """
        task_data = self.__readJsonFile('fixtures/task_data_4.json')
        endpoints = self.exporter.get_app_metrics_endpoints(task_data)
        self.assertEqual(len(endpoints), 0)

    @patch('requests.get')
    def test_get_metrics_1_single_app(self, mock_request):
        """
        Test if app metrics from a single app is returned correctly.
        """
        fixture = self.__readFile('fixtures/app_metrics_1.txt')
        mock_resp = self.__mock_response(content=fixture)
        mock_request.return_value = mock_resp
        response = self.exporter.get_metrics(['http://localhost:6666/metrics'])
        self.assertEqual(response, fixture)

    @patch('requests.get')
    def test_get_metrics_2_multiple_apps(self, mock_request):
        """
        Test if app metrics from multiple apps are returned correctly.
        """
        fixtures = [self.__readFile('fixtures/app_metrics_2a.txt'), self.__readFile('fixtures/app_metrics_2b.txt')]
        mock_request.side_effect = iter([self.__mock_response(content=fixture) for fixture in fixtures])
        response = self.exporter.get_metrics(['http://localhost:6666/metrics', 'http://localhost:6667/metrics'])
        self.assertEqual(response, self.__readFile('fixtures/app_metrics_2_combined.txt'))

if __name__ == '__main__':
    unittest.main(verbosity=2, failfast=True)


main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
import logging
from os import environ
from six.moves.BaseHTTPServer import HTTPServer
from sys import stdout

from mae.mae import MetricsServer

logging.basicConfig(stream=stdout, level=environ.get('LOG_LEVEL', logging.INFO))

MESOS_SLAVE_ADDRESS = 'localhost'
MESOS_SLAVE_PORT = 5051
APP_PORT = 8888


def main():
    try:
        parser = argparse.ArgumentParser()
        parser.add_argument('app_port', help='Port on which this exporter will run on', default=APP_PORT, type=int)
        parser.add_argument('slave_address', help='Mesos slave address', default=MESOS_SLAVE_ADDRESS)
        parser.add_argument('slave_port', help='Mesos slave port', default=MESOS_SLAVE_PORT, type=int)
        args = parser.parse_args()
        server = HTTPServer(('', args.app_port), MetricsServer(args.slave_address, args.slave_port))
        logging.info('Started Mesos app exporter for Prometheus on port {}'.format(args.app_port))
        server.serve_forever()

    except KeyboardInterrupt:
        logging.warning('Ctrl + C received, shutting down the Mesos app exporter for Prometheus...')
        server.socket.close()

if __name__ == '__main__':
    main()


mae/mae.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Mesos App Exporter collects Prometheus metrics from all running executors on a Mesos slave.
"""

from . import __name__, __version__
import json
import logging
import requests
from six.moves.BaseHTTPServer import BaseHTTPRequestHandler

REQUEST_HEADERS = {'User-Agent': '{0}/{1}'.format(__name__, __version__), 'Connection': 'close'}
MESOS_LABEL_ENABLED = 'prometheus.metrics.enabled'
MESOS_LABEL_PORT_INDEX = 'prometheus.metrics.port_index'
MESOS_LABEL_ENDPOINT = 'prometheus.metrics.endpoint'


class MesosAppExporter:
    """
    MesosAppExporter collects metrics from all running Mesos executors.
    """
    def __init__(self, slave_address='localhost', slave_port=5051):
        # type: (str, int) -> None
        """
        Construct a new 'MesosAppExporter' object.

        :param slave_address: Mesos slave process address.
        :param slave_port: Mesos slave process port.
        """
        self.__slave_address = slave_address
        self.__GET_TASKS_PAYLOAD = {'type': 'GET_TASKS'}
        self.__mesos_slave_state = 'http://{0}:{1}/api/v1/state'.format(slave_address, slave_port)
        logging.info('Initialized Mesos App Metrics collector at {}'.format(self.__mesos_slave_state))

    def get_app_metrics_endpoints(self, slave_state=dict()):
        # type: (Dict) -> List
        """
        Derives every Mesos app's metrics endpoints from slave state data.

        :param slave_state: Mesos slave state data.
        """
        tasks = slave_state.get('get_tasks').get('launched_tasks')

        app_metrics_endpoints = list()
        for index, task in enumerate(tasks):
            labels = task.get('labels', {}).get('labels', {})
            label_dict = dict()
            for label in labels:
                label_dict.update({label.get('key'): label.get('value')})

            if MESOS_LABEL_ENABLED in label_dict:
                prom_metrics_port_index = int(label_dict.get(MESOS_LABEL_PORT_INDEX, 0))
                prom_metrics_endpoint = label_dict.get(MESOS_LABEL_ENDPOINT, '/metrics')
                framework_port = tasks[index].get('discovery').get('ports').get('ports')[prom_metrics_port_index].get('number')
                app_metrics_endpoint = 'http://{0}:{1}{2}'.format(self.__slave_address, framework_port, prom_metrics_endpoint)
                app_metrics_endpoints.append(app_metrics_endpoint)

        logging.debug('Got app metrics endpoints from {} apps!'.format(len(app_metrics_endpoints)))
        return app_metrics_endpoints

    def get_metrics(self, endpoints=list()):
        # type: (List) -> str
        """
        Queries each Mesos app's metrics endpoint and returns the combined metrics.

        :param endpoints: List of app metric endpoints.
        """
        combined_metrics = ''
        for index, endpoint in enumerate(endpoints):
            r = requests.get(endpoint, headers=REQUEST_HEADERS)
            app_metrics = r.text
            r.close()
            combined_metrics = '\n'.join([combined_metrics, app_metrics])
        return combined_metrics.strip('\n')

    def get_mesos_slave_state(self):
        # type: (None) -> str
        """
        Queries Mesos slave daemon for its state.
        """
        r = requests.post(self.__mesos_slave_state, json=self.__GET_TASKS_PAYLOAD, headers=REQUEST_HEADERS)
        task_data = r.json()
        r.close()
        return task_data

    def collect(self):
        # type: (None) -> str
        """
        Collects each app's metrics and returns the combined metrics.
        """
        slave_state = self.get_mesos_slave_state()
        endpoints = self.get_app_metrics_endpoints(slave_state)
        combined_metrics = self.get_metrics(endpoints)
        return combined_metrics


def MetricsServer(mesos_slave_address, mesos_slave_port):
    # type: (str, int) -> BaseHTTPRequestHandler
    """
    Metrics HTTP server that servers app metrics for Prometheus to scrape.

    :param mesos_slave_address: Mesos slave process address.
    :param mesos_slave_port: Mesos slave process port.
    """
    class MetricsHTTPHandler(BaseHTTPRequestHandler, object):
        """
        Metrics HTTP handler factory.
        """
        def __init__(self, *args, **kwargs):
            self.server_version = __name__
            self.sys_version = __version__
            self.__mesos_app_exporter = MesosAppExporter(mesos_slave_address, mesos_slave_port)
            super(MetricsHTTPHandler, self).__init__(*args, **kwargs)

        def log_message(self, format, *args):
            return

        def do_GET(self):
            app_metrics = self.__mesos_app_exporter.collect()
            response_body = app_metrics or ''
            self.send_response(200)
            self.send_header('Content-Type', 'text/plain; charset=utf-8')
            self.end_headers()
            self.wfile.write(response_body.encode())
            return
    return MetricsHTTPHandler


mae/cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
import logging
from os import environ
from six.moves.BaseHTTPServer import HTTPServer
from sys import stdout

from .mae import MetricsServer

logging.basicConfig(stream=stdout, level=environ.get('LOG_LEVEL', logging.INFO))

MESOS_SLAVE_ADDRESS = 'localhost'
MESOS_SLAVE_PORT = 5051
APP_PORT = 8888


def ExporterServer(app_port=APP_PORT, slave_address=MESOS_SLAVE_ADDRESS, slave_port=MESOS_SLAVE_PORT):
    # type: (int, str, int) -> HTTPServer
    """
    Returns the exporter as a HTTPServer instance.

    :param app_port: Port the exporter will listen on.
    :param slave_address: Mesos slave address.
    :param slave_port: Mesos slave port.
    """
    exporter = HTTPServer(('', app_port), MetricsServer(slave_address, slave_port))
    logging.info('Starting Mesos app exporter for Prometheus on port {0} and listening to Mesos slave at {1}:{2}'.format(exporter.server_port, slave_address, slave_port))
    return exporter


def main():
    """
    Runs a Prometheus exporter as a HTTP server.
    The default loging level can be overridden with the `LOG_LEVEL` environment variable.
    """
    try:
        parser = argparse.ArgumentParser()
        parser.add_argument('app_port', help='Port on which this exporter will run on', default=APP_PORT, type=int)
        parser.add_argument('slave_address', help='Mesos slave address', default=MESOS_SLAVE_ADDRESS)
        parser.add_argument('slave_port', help='Mesos slave port', default=MESOS_SLAVE_PORT, type=int)
        args = parser.parse_args()
        server = ExporterServer(**vars(args))
        server.serve_forever()

    except KeyboardInterrupt:
        logging.warning('Ctrl + C received, shutting down the Mesos app exporter for Prometheus...')
        server.shutdown()
        server.socket.close()

if __name__ == '__main__':
    main()
